### ISSUE TITLE
pmie rule + metric change produces unknown regardless of data state

### ISSUE THREAD

**Issue by jerdogma** [performancecopilot/pcp#571]
Hello!

I'm running PCP 4.1.3.  I have an app that is supplying metrics through parfait.  This app will supply different metrics based on app state.

In all cases, the metrics show correctly in pminfo.  I can see and fetch values for them.  mmvdecode also show the metric and instances correctly (or at least as correct as I can determine).

But, I have an issue with pmie in this case.  The working sequence of events:
1) System boots
2) pmie starts with no rule referencing metric mmv.com.singlewire.paginggateway.canReceiveBroadcast.  
3) User configures the java app, parfait exposes an instance metric under mmv.com.singlewire.paginggateway.canReceiveBroadcast.
4) User configures pmie with rule:
rising count_inst (mmv.com.singlewire.ipspeaker.canReceiveBroadcast == 0) >= 1
5) pmie sees the metric and functions as expected.  When an instance under mmv.com.singlewire.paginggateway.canReceiveBroadcast gets set to 1, then later gets set to 0, pmie will notice and take its configured action

But, we get a consistent and reproducible failure upon restarting the system (I'll call this the restart case).
1) System boots
2) pmie starts with a rule referencing mmv.com.singlewire.paginggateway.canReceiveBroadcast.  mmv.com.singlewire.paginggateway.canReceiveBroadcast doesn't exist, so pmie correctly says unknown for output.
3) Java app starts, exports mmv.com.singlewire.paginggateway.canReceiveBroadcast instances.
4) We see pmie re-shape the metrics, but the value that pmie shows for them is always unknown.

In the restart case, we:
* Correctly see the metric showing in pminfo
* Can fetch values for the metric using pmval
* Can get pmie to function correctly by restarting the pmie service

It looks to me like pmie, when it sees the new metric, does not handle the instantiation of that metric correctly, such that when calculations on that reshaped metric occur, they always result in unknown.  

I've turned on:
PCP_DEBUG=pmapi,pdu,pmc,appl0,appl1,appl2
I'm attaching pmie debugs for a couple cases:
1) [pmie-unknowns.log](https://github.com/performancecopilot/pcp/files/2516108/pmie-unknowns.log) is the failure case.  In this case, we can see the rule is read correctly, and that the metric shows as unknown.  If you search for reinitMetric, you can see where pmie first sees the new metric (mmv.com.singlewire.paginggateway.canReceiveBroadcast).  It looks to me like the failing operation is the >=.  Skip to the end of the file, we see an expression tree that includes valid values for ==, for count_inst, but not for 0x808881f0, >=.  I conclude this because valid is 0 for it after the metric re-shaping.
2) [pmie-after-alarm-restart.log](https://github.com/performancecopilot/pcp/files/2516109/pmie-after-alarm-restart.log) is the successful case.  In this case, I restart pmie with the same rules present but with the application exposing the metrics, so no reshaping is required.

I'm about to venture into wild speculation into code that I only started reading yesterday, so I may be way off base.

reinitMetric: re-shaped ...
Expr dump @ 0x808881f0
  op=35 (>=) arg1=0x808880d0 arg2=0x80888130 parent=0x80888250
  eval=cndGte_n_1 metrics=0x80887dc0 ring=0x808841a0
  valid=0 cardinality[H,I,T]=[1,2,1] tspan=2
  nsmpls=2 nvals=4 sem=TRUTH units=

I think the tspan value is incorrect here, it should be 1 and not 2.  But, you all are the experts here, not me.

Any help would be very much appreciated.  Thanks!

I'd like to avoid restarting pmie when new metrics get exposed by the java app.  I've looked for pmie switches or configuration that might help here, and I'm coming up empty.

**Comment by kmcdonell** [comment]
@jerdogma I'd really wish Seppo Keronen would jump in here, ... but that's not going to happen, so I guess this one ends on my plate.  Thanks for the detailed bug report, attached logs and triage ... I'll let you know what I discover.

**Comment by jerdogma** [comment]
@kmcdonell Understood.  In the restart/failure case, the number of instances will remain the same at 2.  Our expectations:
* No instances for 2 consecutive intervals: rising count_inst(..) >= 1 false
* 1 instance, where metric instance is true for 2 consecutive intervals: rising count_inst(..) >= 1 false
* 2 instances, where metric instances are both true for 2 consecutive intervals: rising count_inst(..) >= 1 false
* 2 instances, where both metric instances are true for one interval, then one metric instance is false and one true: rising count_inst(..) >= 1 true
* N instances, where N metric instances are true for one interval, then in the last interval one metric instance is false and N-1 are true: rising count_inst(..) >= 1 true

I hope I have the logic and expectations of pmie right.  I read the manual and wrote the rule, but then I studied the code and debugs and learned that what pmie is doing is not at all simple.  Rising appears to require two consecutive intervals (makes sense given it's supposed to capture change) and it's not getting that in the failure case, so it results in unknown.  It's at that point I can't quite figure out what's going on.  It's hard for me to build a mental model that matches what pmie is doing, to match instances and values with the code.  I see that the metric instances evaluate to true, false, which tells me that the metrics themselves are present and being accessed correctly.  

Basically, we're trying to capture the time when the instance counts of metrics are changing state.  These instances represent devices in the network and the metric represents their state.  They are going from a registered to an unregistered state, and pmie is alerting the admin.

**Comment by kmcdonell** [comment]
@jerdogma reports that the when the fix is applied to their environment the issue is resolved.
